File: templates/python/minimal_provider_fallback.py
"""
Minimal Provider Router — Hera Architecture Reference (Python)

Inspired by 9router (decolua/9router) — production AI router with multi-key pool,
round-robin with sticky limit, account-level cooldown, exponential backoff,
and config-driven error classification.

Provides:
- Multi-key pool per provider
- "fallback" strategy (sequential) or "round-robin" with sticky limit
- Account-level cooldown (rateLimitedUntil ISO timestamp)
- Exponential backoff (2s -> 4s -> 8s, max 5 min)
- Per-model lock (separate from account cooldown)
- Config-driven error rules (text + status matching)
"""

from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass, field, replace
from typing import Any, Awaitable, Callable, Literal

# ============================================================================
# Types
# ============================================================================

Strategy = Literal["fallback", "round-robin"]


@dataclass
class ApiKey:
    id: str
    key: str
    rate_limited_until: str | None = None  # ISO timestamp
    backoff_level: int = 0
    last_error: dict | None = None


@dataclass
class ErrorRule:
    text: str | None = None
    status: int | None = None
    cooldown_ms: int | None = None
    backoff: bool = False


@dataclass
class RouterConfig:
    strategy: Strategy = "fallback"
    sticky_limit: int = 1
    max_attempts: int = 3
    backoff_ms: int = 2000
    backoff_max_ms: int = 5 * 60 * 1000
    transient_cooldown_ms: int = 30_000
    max_backoff_level: int = 15


# Default error rules (text first, then status)
DEFAULT_ERROR_RULES: list[ErrorRule] = [
    ErrorRule(text="no credentials", cooldown_ms=2 * 60 * 1000),
    ErrorRule(text="improperly formed request", cooldown_ms=2 * 60 * 1000),
    ErrorRule(text="rate limit", backoff=True),
    ErrorRule(text="too many requests", backoff=True),
    ErrorRule(text="quota exceeded", backoff=True),
    ErrorRule(text="capacity", backoff=True),
    ErrorRule(text="overloaded", backoff=True),
    ErrorRule(status=401, cooldown_ms=2 * 60 * 1000),
    ErrorRule(status=403, cooldown_ms=2 * 60 * 1000),
    ErrorRule(status=404, cooldown_ms=2 * 60 * 1000),
    ErrorRule(status=429, backoff=True),
]

NON_FALLBACK_STATUSES = {400}


# ============================================================================
# Error classification
# ============================================================================

def compute_backoff(level: int, base: int, max_ms: int) -> int:
    """Exponential backoff: level=1 -> base, level=2 -> base*2, ..."""
    return min(base * 2 ** max(0, level - 1), max_ms)


def check_fallback_error(
    status: int,
    error_text: str,
    backoff_level: int,
    config: RouterConfig,
    rules: list[ErrorRule] | None = None,
) -> dict:
    """Classify error: should we try next key? How long to cooldown?"""
    if status in NON_FALLBACK_STATUSES:
        return {"should_fallback": False, "cooldown_ms": 0}

    rules = rules or DEFAULT_ERROR_RULES
    lower = (error_text or "").lower()

    for rule in rules:
        # Text match
        if rule.text and rule.text in lower:
            if rule.backoff:
                new_level = min(backoff_level + 1, config.max_backoff_level)
                return {
                    "should_fallback": True,
                    "cooldown_ms": compute_backoff(new_level, config.backoff_ms, config.backoff_max_ms),
                    "new_backoff_level": new_level,
                }
            return {
                "should_fallback": True,
                "cooldown_ms": rule.cooldown_ms or config.transient_cooldown_ms,
            }
        # Status match
        if rule.status == status:
            if rule.backoff:
                new_level = min(backoff_level + 1, config.max_backoff_level)
                return {
                    "should_fallback": True,
                    "cooldown_ms": compute_backoff(new_level, config.backoff_ms, config.backoff_max_ms),
                    "new_backoff_level": new_level,
                }
            return {
                "should_fallback": True,
                "cooldown_ms": rule.cooldown_ms or config.transient_cooldown_ms,
            }

    # Default: transient cooldown
    return {"should_fallback": True, "cooldown_ms": config.transient_cooldown_ms}


# ============================================================================
# Key state
# ============================================================================

def is_key_available(key: ApiKey) -> bool:
    if not key.rate_limited_until:
        return True
    return time.time() >= _iso_to_ts(key.rate_limited_until)


def filter_available_keys(keys: list[ApiKey]) -> list[ApiKey]:
    return [k for k in keys if is_key_available(k)]


def _iso_to_ts(iso: str) -> float:
    """Parse ISO timestamp to epoch seconds (works without external deps)."""
    from datetime import datetime
    return datetime.fromisoformat(iso.replace("Z", "+00:00")).timestamp()


def _ts_to_iso(ts: float) -> str:
    from datetime import datetime, timezone
    return datetime.fromtimestamp(ts, tz=timezone.utc).isoformat()


def apply_error_state(key: ApiKey, status: int, error_text: str, config: RouterConfig) -> ApiKey:
    """Update key with error state and appropriate cooldown."""
    result = check_fallback_error(status, error_text, key.backoff_level, config)
    cooldown_ms = result["cooldown_ms"]
    return replace(
        key,
        rate_limited_until=(_ts_to_iso(time.time() + cooldown_ms / 1000) if cooldown_ms > 0 else None),
        backoff_level=result.get("new_backoff_level", key.backoff_level),
        last_error={"status": status, "message": error_text, "timestamp": _ts_to_iso(time.time())},
    )


def reset_key_state(key: ApiKey) -> ApiKey:
    """Clear cooldown and reset backoff (call after success)."""
    return replace(key, rate_limited_until=None, backoff_level=0, last_error=None)


# ============================================================================
# Round-robin state (module-level, keyed by route name)
# ============================================================================

@dataclass
class RotationState:
    index: int = 0
    consecutive_use_count: int = 0


_sticky_states: dict[str, RotationState] = {}


def _pick_next_key(
    keys: list[ApiKey], state_key: str, strategy: Strategy, sticky_limit: int
) -> tuple[ApiKey, RotationState] | None:
    available = filter_available_keys(keys)
    if not available:
        return None

    if strategy == "fallback" or len(available) == 1:
        return available[0], RotationState()

    # round-robin
    state = _sticky_states.get(state_key, RotationState())
    selected_idx = state.index % len(keys)

    # If current key is in cooldown, find next available
    if not is_key_available(keys[selected_idx]):
        for i in range(len(keys)):
            idx = (state.index + i) % len(keys)
            if is_key_available(keys[idx]):
                selected_idx = idx
                break

    new_consecutive = state.consecutive_use_count + 1
    next_state = (
        RotationState(index=(selected_idx + 1) % len(keys))
        if new_consecutive >= sticky_limit
        else RotationState(index=selected_idx, consecutive_use_count=new_consecutive)
    )
    return keys[selected_idx], next_state


# ============================================================================
# ProviderRouter
# ============================================================================

ProviderCall = Callable[[dict, ApiKey], Awaitable[dict]]


class ProviderRouter:
    """Routes LLM calls across multiple API keys with rotation + cooldown."""

    def __init__(
        self,
        route_name: str,
        keys: list[ApiKey],
        provider_call: ProviderCall,
        config: RouterConfig | None = None,
    ):
        if not keys:
            raise ValueError("ProviderRouter requires at least 1 key")
        self.route_name = route_name
        self.keys = keys
        self.provider_call = provider_call
        self.config = config or RouterConfig()

    def set_keys(self, keys: list[ApiKey]) -> None:
        if not keys:
            raise ValueError("Need at least 1 key")
        self.keys = keys

    def get_key_states(self) -> list[dict]:
        return [
            {"id": k.id, "available": is_key_available(k), "last_error": k.last_error}
            for k in self.keys
        ]

    def reset_all(self) -> None:
        self.keys = [reset_key_state(k) for k in self.keys]
        _sticky_states.clear()

    async def call(self, ctx: dict) -> dict:
        last_error: Exception | None = None

        for attempt in range(1, self.config.max_attempts + 1):
            pick = _pick_next_key(
                self.keys, self.route_name, self.config.strategy, self.config.sticky_limit
            )
            if pick is None:
                raise RuntimeError(
                    f"No available keys for route '{self.route_name}' (all in cooldown)"
                )

            selected_key, next_state = pick
            _sticky_states[self.route_name] = next_state

            try:
                result = await self.provider_call(ctx, selected_key)
                # Success: reset this key
                self.keys = [
                    reset_key_state(k) if k.id == selected_key.id else k
                    for k in self.keys
                ]
                return result
            except Exception as err:
                last_error = err
                status = getattr(err, "status", None) or 500
                message = str(err)

                # Apply error state to this key
                self.keys = [
                    apply_error_state(k, status, message, self.config)
                    if k.id == selected_key.id
                    else k
                    for k in self.keys
                ]

                classification = check_fallback_error(
                    status, message, selected_key.backoff_level, self.config
                )
                if not classification["should_fallback"]:
                    raise

                if attempt >= self.config.max_attempts:
                    raise RuntimeError(
                        f"All {self.config.max_attempts} attempts failed for "
                        f"'{self.route_name}'. Last: {message}"
                    ) from err

        raise last_error or RuntimeError("Unknown error in ProviderRouter.call")

File: templates/minimal-provider-fallback.ts
/**
 * Minimal Provider Router — Hera Architecture Reference (TypeScript)
 *
 * Inspired by 9router (decolua/9router) — a production AI router with:
 * - Multi-key pool per provider
 * - Round-robin strategy (with sticky limit)
 * - Account-level cooldown (skip rate-limited accounts)
 * - Exponential backoff per key
 * - Per-model lock (separate from account cooldown)
 * - Config-driven error classification
 *
 * Based on:
 * - 9router: open-sse/services/combo.js
 * - 9router: open-sse/services/accountFallback.js
 * - 9router: open-sse/config/errorConfig.js
 *
 * Provides two strategies:
 * - "fallback"   — try primary, then secondary, etc. (sequential)
 * - "round-robin" — rotate with optional sticky limit (N requests per key)
 */

// ============================================================================
// Inline types
// ============================================================================

interface AgentContext {
  systemPrompt: string;
  messages: Array<{ role: string; content: unknown }>;
}

interface ContentBlock {
  type: "text" | "toolCall";
  text?: string;
  id?: string;
  name?: string;
  arguments?: unknown;
}

interface CallResult {
  content: ContentBlock[];
  usage?: { inputTokens: number; outputTokens: number };
}

/** A single API key/account, with its current cooldown state. */
export interface ApiKey {
  id: string;
  key: string;
  rateLimitedUntil?: string;     // ISO timestamp; if in future, skip this key
  backoffLevel?: number;          // For exponential backoff (0, 1, 2, ...)
  lastError?: { status: number; message: string; timestamp: string };
}

/** Strategy for picking a key. */
export type Strategy = "fallback" | "round-robin";

export interface RouterConfig {
  /** Pool of API keys for this provider. (Optional here; ProviderRouter constructor takes keys as a separate param.) */
  keys?: ApiKey[];
  /** Strategy (default: "fallback"). */
  strategy?: Strategy;
  /**
   * Sticky limit for round-robin — number of consecutive requests
   * to use the same key before rotating. Default 1.
   * Set to N to "stick" to the same key for N requests.
   */
  stickyLimit?: number;
  /** Max attempts before giving up (default: 3). */
  maxAttempts?: number;
  /** Backoff base in ms (default: 2000). */
  backoffMs?: number;
  /** Backoff max in ms (default: 5 * 60 * 1000 = 5 minutes). */
  backoffMaxMs?: number;
  /** Cooldown for transient/unknown errors in ms (default: 30_000). */
  transientCooldownMs?: number;
  /** Optional per-model lock TTL — locks a key for a specific model. */
  modelLockTtlMs?: number;
}

// ============================================================================
// Error classification
// ============================================================================

/** Error classification rule. Checked top-to-bottom. */
export interface ErrorRule {
  /** Substring match in error text (case-insensitive). */
  text?: string;
  /** HTTP status code to match. */
  status?: number;
  /** Fixed cooldown in ms. */
  cooldownMs?: number;
  /** If true, use exponential backoff instead of fixed cooldown. */
  backoff?: boolean;
}

/**
 * Default error rules (from 9router's errorConfig.js).
 * Order matters — text rules checked first, then status.
 */
export const DEFAULT_ERROR_RULES: ErrorRule[] = [
  // Text-based
  { text: "no credentials",          cooldownMs: 2 * 60 * 1000 },
  { text: "improperly formed request", cooldownMs: 2 * 60 * 1000 },
  { text: "rate limit",              backoff: true },
  { text: "too many requests",       backoff: true },
  { text: "quota exceeded",          backoff: true },
  { text: "capacity",                backoff: true },
  { text: "overloaded",              backoff: true },
  // Status-based
  { status: 401, cooldownMs: 2 * 60 * 1000 },
  { status: 403, cooldownMs: 2 * 60 * 1000 },
  { status: 404, cooldownMs: 2 * 60 * 1000 },
  { status: 429, backoff: true },
];

/** Cooldown for non-fallback (terminal) errors. */
const NON_FALLBACK_STATUSES = new Set([400]);

export function checkFallbackError(
  status: number,
  errorText: string,
  backoffLevel: number,
  rules: ErrorRule[] = DEFAULT_ERROR_RULES,
  backoffBase = 2000,
  backoffMax = 5 * 60 * 1000,
  transientCooldown = 30_000,
  maxLevel = 15
): { shouldFallback: boolean; cooldownMs: number; newBackoffLevel?: number } {
  // Terminal errors — don't fall back
  if (NON_FALLBACK_STATUSES.has(status)) {
    return { shouldFallback: false, cooldownMs: 0 };
  }

  const lower = (errorText || "").toLowerCase();

  for (const rule of rules) {
    // Text match (case-insensitive)
    if (rule.text && lower.includes(rule.text)) {
      if (rule.backoff) {
        const newLevel = Math.min(backoffLevel + 1, maxLevel);
        return {
          shouldFallback: true,
          cooldownMs: computeBackoff(newLevel, backoffBase, backoffMax),
          newBackoffLevel: newLevel,
        };
      }
      return { shouldFallback: true, cooldownMs: rule.cooldownMs ?? transientCooldown };
    }
    // Status match
    if (rule.status === status) {
      if (rule.backoff) {
        const newLevel = Math.min(backoffLevel + 1, maxLevel);
        return {
          shouldFallback: true,
          cooldownMs: computeBackoff(newLevel, backoffBase, backoffMax),
          newBackoffLevel: newLevel,
        };
      }
      return { shouldFallback: true, cooldownMs: rule.cooldownMs ?? transientCooldown };
    }
  }

  // Default: transient cooldown
  return { shouldFallback: true, cooldownMs: transientCooldown };
}

function computeBackoff(level: number, base: number, max: number): number {
  // level=1 → base, level=2 → base*2, level=3 → base*4, ...
  const ms = base * 2 ** Math.max(0, level - 1);
  return Math.min(ms, max);
}

// ============================================================================
// Per-key state
// ============================================================================

function isKeyAvailable(key: ApiKey): boolean {
  if (!key.rateLimitedUntil) return true;
  return new Date(key.rateLimitedUntil).getTime() <= Date.now();
}

/** Filter out keys in cooldown. */
export function filterAvailableKeys(keys: ApiKey[]): ApiKey[] {
  return keys.filter(isKeyAvailable);
}

function applyErrorState(
  key: ApiKey,
  status: number,
  errorText: string,
  config: Required<RouterConfig>
): ApiKey {
  const backoffLevel = key.backoffLevel ?? 0;
  const { cooldownMs, newBackoffLevel } = checkFallbackError(
    status,
    errorText,
    backoffLevel,
    DEFAULT_ERROR_RULES,
    config.backoffMs,
    config.backoffMaxMs,
    config.transientCooldownMs
  );

  return {
    ...key,
    rateLimitedUntil: cooldownMs > 0
      ? new Date(Date.now() + cooldownMs).toISOString()
      : undefined,
    backoffLevel: newBackoffLevel ?? backoffLevel,
    lastError: { status, message: errorText, timestamp: new Date().toISOString() },
  };
}

function resetKeyState(key: ApiKey): ApiKey {
  return {
    ...key,
    rateLimitedUntil: undefined,
    backoffLevel: 0,
    lastError: undefined,
  };
}

// ============================================================================
// Round-robin state (per RouterConfig instance, keyed by route name)
// ============================================================================

/** Track rotation state: how many requests used current key, and which key index. */
interface RotationState {
  index: number;
  consecutiveUseCount: number;
}

/** Sticky state — same key for N consecutive requests. */
const stickyStates = new Map<string, RotationState>();

/** Get next key based on strategy. Returns null if all keys unavailable. */
function pickNextKey(
  keys: ApiKey[],
  stateKey: string,
  strategy: Strategy,
  stickyLimit: number
): { key: ApiKey; nextState: RotationState } | null {
  const available = filterAvailableKeys(keys);
  if (available.length === 0) return null;

  if (strategy === "fallback" || available.length === 1) {
    // Always pick the first available
    return { key: available[0], nextState: { index: 0, consecutiveUseCount: 0 } };
  }

  // round-robin with sticky limit
  const state = stickyStates.get(stateKey) ?? { index: 0, consecutiveUseCount: 0 };
  const currentKey = keys[state.index % keys.length];

  // If current key is still in cooldown, find next available
  let selectedIdx = state.index % keys.length;
  if (!isKeyAvailable(currentKey)) {
    // Find next available starting from current index
    for (let i = 0; i < keys.length; i++) {
      const idx = (state.index + i) % keys.length;
      if (isKeyAvailable(keys[idx])) {
        selectedIdx = idx;
        break;
      }
    }
  }

  const newConsecutive = state.consecutiveUseCount + 1;
  const nextState: RotationState =
    newConsecutive >= stickyLimit
      ? { index: (selectedIdx + 1) % keys.length, consecutiveUseCount: 0 }
      : { index: selectedIdx, consecutiveUseCount: newConsecutive };

  return { key: keys[selectedIdx], nextState };
}

// ============================================================================
// Provider interface (must be implemented by your actual provider)
// ============================================================================

export interface ProviderCall {
  (ctx: AgentContext, key: ApiKey): Promise<CallResult>;
}

// ============================================================================
// ProviderRouter — main class
// ============================================================================

export class ProviderRouter {
  private keys: ApiKey[];
  private config: Required<RouterConfig>;
  private providerCall: ProviderCall;
  private routeName: string;

  constructor(
    routeName: string,
    keys: ApiKey[],
    providerCall: ProviderCall,
    config: RouterConfig = {}
  ) {
    if (keys.length === 0) throw new Error("ProviderRouter requires at least 1 key");
    this.routeName = routeName;
    this.keys = keys;
    this.providerCall = providerCall;
    this.config = {
      keys,
      strategy: config.strategy ?? "fallback",
      stickyLimit: config.stickyLimit ?? 1,
      maxAttempts: config.maxAttempts ?? 3,
      backoffMs: config.backoffMs ?? 2000,
      backoffMaxMs: config.backoffMaxMs ?? 5 * 60 * 1000,
      transientCooldownMs: config.transientCooldownMs ?? 30_000,
      modelLockTtlMs: config.modelLockTtlMs ?? 0,
    };
  }

  /** Update the key pool (e.g., when user adds/removes a key). */
  setKeys(keys: ApiKey[]): void {
    if (keys.length === 0) throw new Error("Need at least 1 key");
    this.keys = keys;
    this.config.keys = keys;
  }

  /** Get current key pool state (for UI/observability). */
  getKeyStates(): Array<{ id: string; available: boolean; lastError?: ApiKey["lastError"] }> {
    return this.keys.map((k) => ({
      id: k.id,
      available: isKeyAvailable(k),
      lastError: k.lastError,
    }));
  }

  /** Reset all key states (e.g., on app restart). */
  resetAll(): void {
    this.keys = this.keys.map(resetKeyState);
    stickyStates.clear();
  }

  /** Make a call. Tries keys in order (fallback) or rotates (round-robin). */
  async call(ctx: AgentContext): Promise<CallResult> {
    const { strategy, stickyLimit, maxAttempts } = this.config;
    let attempt = 0;
    let lastError: unknown;

    while (attempt < maxAttempts) {
      attempt++;

      const pick = pickNextKey(this.keys, this.routeName, strategy, stickyLimit);
      if (!pick) {
        throw new Error(`No available keys for route "${this.routeName}" (all in cooldown)`);
      }

      // Save rotation state
      stickyStates.set(this.routeName, pick.nextState);

      try {
        const result = await this.providerCall(ctx, pick.key);

        // Success: reset this key's state
        this.keys = this.keys.map((k) => (k.id === pick.key.id ? resetKeyState(k) : k));

        return result;
      } catch (err: any) {
        lastError = err;
        const status = err?.status ?? err?.response?.status ?? 500;
        const message = err?.message ?? String(err);

        // Apply error state to this key
        this.keys = this.keys.map((k) =>
          k.id === pick.key.id ? applyErrorState(k, status, message, this.config) : k
        );

        const { shouldFallback } = checkFallbackError(
          status,
          message,
          pick.key.backoffLevel ?? 0,
          DEFAULT_ERROR_RULES,
          this.config.backoffMs,
          this.config.backoffMaxMs,
          this.config.transientCooldownMs
        );

        if (!shouldFallback) {
          throw err;  // Terminal — don't try other keys
        }

        // If last attempt, throw
        if (attempt >= maxAttempts) {
          throw new Error(
            `All ${maxAttempts} attempts failed for "${this.routeName}". Last: ${message}`
          );
        }
      }
    }

    throw lastError instanceof Error ? lastError : new Error(String(lastError));
  }
}

// ============================================================================
// Example usage
// ============================================================================

// // 1. Define a provider call function (your actual API client)
// const callOpenAI: ProviderCall = async (ctx, key) => {
//   const response = await fetch("https://api.openai.com/v1/chat/completions", {
//     method: "POST",
//     headers: { Authorization: `Bearer ${key.key}`, "Content-Type": "application/json" },
//     body: JSON.stringify({ model: "gpt-4o-mini", messages: ctx.messages }),
//   });
//   if (!response.ok) throw Object.assign(new Error(await response.text()), { status: response.status });
//   const data = await response.json();
//   return { content: [{ type: "text", text: data.choices[0].message.content }] };
// };
//
// // 2. Create router with 3 keys, round-robin with stickyLimit=2
// const router = new ProviderRouter("openai-gpt4o", [
//   { id: "key-1", key: process.env.OPENAI_KEY_1! },
//   { id: "key-2", key: process.env.OPENAI_KEY_2! },
//   { id: "key-3", key: process.env.OPENAI_KEY_3! },
// ], callOpenAI, { strategy: "round-robin", stickyLimit: 2 });
//
// // 3. Make calls
// const result = await router.call({ systemPrompt: "...", messages: [...] });
//
// // 4. Check key states
// console.log(router.getKeyStates());
